source/openapi-admin-v3.yaml

```diff
@@ -5132,55 +5178,517 @@ components:
         - name
         - function_source
         - respond_result
-    UserProvider:
-      type: object
-      properties:
-        id:
-          type: string
-        provider_type:
-          "$ref": "#/components/schemas/ProviderType"
-        provider_id:
-          type: string
-    ProviderSummary:
+    
+
+
+
+
+
+
+
+
+
+    BaseAuthProvider:
+      required:
+        - name
+        - type
       properties:
         _id:
           type: string
+          description: The authentication provider's unique ID.
         name:
           type: string
           description: The name of the authentication provider. This value must be the same as the value of `type`.
         type:
           "$ref": "#/components/schemas/ProviderType"
         disabled:
           type: boolean
-    FullProvider:
+          description: If `true`, the auth provider is disabled. Users cannot log in through the provider until it is enabled.
+    
+    AuthProvider:
+      anyOf:
+        - $ref: "#/components/schemas/AnonymousAuthProvider"
+        - $ref: "#/components/schemas/EmailPasswordAuthProvider"
+        - $ref: "#/components/schemas/ApiKeyAuthProvider"
+        - $ref: "#/components/schemas/CustomJwtAuthProvider"
+        - $ref: "#/components/schemas/CustomFunctionAuthProvider"
+        - $ref: "#/components/schemas/AppleAuthProvider"
+        - $ref: "#/components/schemas/GoogleAuthProvider"
+        - $ref: "#/components/schemas/FacebookAuthProvider"
+    
+    AnonymousAuthProvider:
+      allOf: 
+        - $ref: "#/components/schemas/BaseAuthProvider"
+        - properties:
```

Comment:
nice

Reply:
Nice _except_ Redoc does not play nicely with these composed types so there's a few warts like this:

- `title` resolves to an intermediate schema name instead of the actual schema name. So e.g. `GoogleAuthProvider` renders as `NonAppleAuthProvider`.
   
   ![Screen Shot 2022-11-21 at 2 54 54 PM](https://user-images.githubusercontent.com/15657698/203146889-9c057cd6-7af1-461c-a4d2-39b93d246bb3.png)

- An `allOf` that combines multiple `oneOf` schemas only ends up resolving to the last one in the list. That's why we have types like
   ```
   config:
     oneOf:
       - allOf:
         - title: ConfirmationAutomaticRefreshBuiltIn
         - $ref: "#/components/schemas/EmailConfirmationAutomatic"
         - $ref: "#/components/schemas/PasswordRefreshBuiltIn"
       - allOf:
         - title: ConfirmationAutomaticRefreshCustom
         - $ref: "#/components/schemas/EmailConfirmationAutomatic"
         - $ref: "#/components/schemas/PasswordRefreshCustomFunction"
       - ...
   ```
   
   instead of
   
   ```
   config:
     allOf:
       - $ref: "#/components/EmailConfirmationType"
       - $ref: "#/components/PasswordResetType"
   ```

(this reminds me that I also need to delete `EmailConfirmationType` and `PasswordResetType` because they aren't used anymore!)